02-peer-protocol.md

```diff
@@ -560,6 +1077,337 @@ ones it has sent so it can use them should they be requested by
+Funding composition for channel establishment v2 makes use of the
+[Interactive Transaction Construction](#interactive-transaction-construction)
+protocol, with the following additional caveats.
+
+#### The `tx_add_input` Message
+
+##### Requirements
+
+The sending node:
+  - if the receiver set `require_confirmed_inputs` in `open_channel2` or `accept_channel2`:
+    - MUST NOT send a `tx_add_input` that contains an unconfirmed input
+
+#### The `tx_add_output` Message
+
+##### Requirements
+
+The sending node:
+  - if is the *opener*:
+    - MUST send at least one `tx_add_output`,  which contains the
+      channel's funding output
+
+##### Rationale
+
+The channel funding output must be added by the *opener*, who pays its fees.
+
+#### The `tx_complete` Message
+
+Upon receipt of consecutive `tx_complete`s, the receiving node:
+  - if is the *accepter*:
+    - MUST fail the negotiation if:
+      - no funding output was received
+      - the value of the funding output is not equal to the sum of
+        `open_channel2`.`funding_satoshis` and `accept_channel2`.
+        `funding_satoshis`
+      - the value of the funding output is less than the `dust_limit`
+  - if it is an RBF attempt:
+    - MUST fail the negotiation if:
+      - the transaction's total fees is less than the last
+        successfully negotiated transaction's fees
+      - the transaction does not share a common input with all previous
+        funding transactions
+  - if it has sent `require_confirmed_inputs` in `open_channel2` or `accept_channel2`:
+    - MUST fail the negotiation if:
+      - one of the inputs added by the other peer is unconfirmed
+
+### The `commitment_signed` Message
+
+This message is exchanged by both peers. It contains the signatures for
+the first commitment transaction.
+
+Rationale and Requirements are the same as listed below,
+for [`commitment_signed`](#commiting-updates-so-far-commitment_signed) with the following additions.
+
+#### Requirements
+
+The sending node:
+  - MUST send zero HTLCs.
+
+The receiving node:
+  - if the message has one or more HTLCs:
+    - MUST fail the negotiation
+  - if it has not already transmitted its `commitment_signed`:
+    - MUST send `commitment_signed`
+  - Otherwise:
+    - MUST send `tx_signatures` if it should sign first, as specified
+      in the [`tx_signatures` requirements](#the-tx_signatures-message)
+
+#### Rationale
+
+The first commitment transaction has no HTLCs.
+
+### Sharing funding signatures: `tx_signatures`
+
+After a valid `commitment_signed` has been received
+from the peer and a `commitment_signed` has been sent, a peer:
+  - MUST transmit `tx_signatures` with their signatures for the funding
+    transaction, following the order specified in the
+    [`tx_signatures` requirements](#the-tx_signatures-message)
+
+#### Requirements
+
+The sending node:
+  - MUST verify it has received a valid commitment signature from its peer
+  - MUST remember the details of this funding transaction
```

Comment:
I think we should move that requirement to the sending side of `commitment_signed`. Combined with an additional mechanism to retransmit `commitment_signed` on reconnection when necessary, this gets rid of cases where one side has stored the channel state and the other hasn't (in which case we would previously need to actively double-spend the channel or wait for the other participant to actively double-spend it, which is annoying). I think that we may not even need to change anything for the reestablish logic, we can probably distinguish the case where `next_commitment_number` is `0` from the one where it is `1` to figure out whether the first `commitment_signed` needs to be retransmitted or not.

I provided more details in [this gist](https://gist.github.com/t-bast/2594ecf0314f0b91293a3f5e3bf10540), with diagrams of the various scenarios that create issues.

Reply:
Here is a more detailed proposal, after doing some implementation work in eclair.

The first step is to move the storage requirement to the `commitment_signed` section: once `tx_complete` has been exchanged, we store the transaction details and send `commitment_signed`.
There is no need for an ordering on `commitment_signed`: both peers should send it as soon as `tx_complete` has been exchanged.

If a disconnection happens before `tx_signatures` has been received by both peers, we resume the signing session when reconnecting.
In that case, at least one of the peers is missing either:

- `commitment_signed` and `tx_signatures`
- only `tx_signatures` if they're already received `commitment_signed`

We introduce a new tlv to the `channel_reestablish` message to ask our peer to retransmit the missing messages.
This tlv must not be included if we've already received `commitment_signed` and `tx_signatures`.

1. `tlv_stream`: `channel_reestablish_tlvs`
2. types:
  1. type: 0 (`next_funding`)
  2. data:
    * [`sha256`:`funding_txid`]
    * [`byte`:`retransmit_commitment_signed`]

We can get rid of the existing requirement of always retransmitting `tx_signatures` now that we explicitly ask our peer to retransmit it if necessary.

A node that receives `channel_reestablish` that contains the `next_funding` tlv:

- If it doesn't have a channel for that `channel_id`:
  - MUST send an `error` and close the connection.
  - Note that the fact that the node didn't remember that channel means it never sent its `commitment_signed`, which means the other node hasn't sent its `tx_signatures` and can thus safely forget that channel when receiving the `error`.
- If `funding_txid` doesn't match one of the funding transactions:
  - MUST send `tx_abort` and resume the channel.
  - This must be for an RBF attempt where we got disconnected before storing the RBF transaction, which means we haven't sent our `commitment_signed`.
  - This `tx_abort` lets the other peer know that they can safely forget that RBF attempt (since we didn't send `commitment_signed`, they haven't sent `tx_signatures` so they're not risking anything).
- If `funding_txid` matches one of the funding transactions:
  - If `retransmit_commitment_signed` is set to `0x01`, MUST retransmit `commitment_signed` and `tx_signatures`.
  - If `retransmit_commitment_signed` is set to `0x00`, MUST retransmit `tx_signatures`.

With those changes, the following scenarios are nicely handled (variations of those scenarios are left as an exercise to the reader :D).
The corresponding scenarios will work for splicing as well, where it's crucial to correctly handle them (otherwise they would lead to a unilateral close).

## Half-forgotten channel

```text
  Alice                          Bob
    |        tx_complete          |
    |<----------------------------|
    |        tx_complete          |
    |------------->X              |
    |        commit_sig           |
    |------------->X              |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |           error             |
    |<----------------------------|
```

Alice's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`.
Bob has forgotten that channel and sends an `error`: when receiving that `error`, Alice can simply forget that channel too.

## Half-forgotten RBF attempt

```text
  Alice                          Bob
    |        tx_complete          |
    |---------------------------->|
    |        tx_complete          |
    |<----------------------------|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
    |        tx_init_rbf          |
    |---------------------------->|
    |        tx_ack_rbf           |
    |---------------------------->|
    |           ...               |
    |        tx_complete          |
    |<----------------------------|
    |        tx_complete          |
    |------------->X              |
    |        commit_sig           |
    |------------->X              |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |    channel_reestablish      |
    |<----------------------------|
    |          tx_abort           |
    |<----------------------------|
    |          tx_abort           |
    |---------------------------->|
```

Alice's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`, using the `funding_txid` of the RBF transaction.
Bob has forgotten that RBF attempt on disconnection, so its `channel_reestablish` doesn't contain `next_funding`.
Alice's `next_funding.funding_txid` doesn't match anything on Bob's side, so Bob sends `tx_abort`.
Alice receives that `tx_abort`, cancels the RBF attempt (which is safe since she hasn't sent `tx_signatures`) and echoes `tx_abort`.
Both sides have correctly cleaned up that failed RBF attempt.

## Partially signed channel

```text
  Alice                          Bob
    |        tx_complete          |
    |---------------------------->|
    |        tx_complete          |
    |<----------------------------|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |--------------->X            |
    |       tx_signatures         |
    |--------------->X            |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |    channel_reestablish      |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
```

In this example, Alice was supposed to send `tx_signatures` first according to the ordering requirements.
Alice's `channel_reestablish` contains `next_funding` without `retransmit_commitment_signed`.
Bob's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`.
Alice retransmits `commitment_signed` and `tx_signatures`, and Bob sends his `tx_signatures` once he has received Alice's.
The channel is now fully functional.

## Partially signed RBF attempt

```text
  Alice                          Bob
    |        tx_complete          |
    |---------------------------->|
    |        tx_complete          |
    |<----------------------------|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
    |        tx_init_rbf          |
    |---------------------------->|
    |        tx_ack_rbf           |
    |---------------------------->|
    |           ...               |
    |        tx_complete          |
    |<----------------------------|
    |        tx_complete          |
    |---------------------------->|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |--------------->X            |
    |       tx_signatures         |
    |--------------->X            |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |    channel_reestablish      |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
```

In this example, Alice was supposed to send `tx_signatures` first according to the ordering requirements.
Alice's `channel_reestablish` contains `next_funding` without `retransmit_commitment_signed`, using the `funding_txid` of the RBF attempt.
Bob's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`, using the `funding_txid` of the RBF attempt.
Alice retransmits `commitment_signed` and `tx_signatures`, and Bob sends his `tx_signatures` once he has received Alice's.
This completes the RBF attempt.

@niftynei @ddustin @rustyrussell I believe that this is also a better way of handling disconnection/reestablish for splicing, compared to incrementing the `commitment_index` like you currently proposed (which requires re-signing all active commitments and revoking the previous one for no added benefits: see https://github.com/lightning/bolts/pull/863#discussion_r1135122533 for the corresponding splice discussion).

Reply:
We can even simplify this by getting rid of the `retransmit_commitment_signed` field, and whenever the `next_funding` tlv is present, always retransmit both `commitment_signed` and `tx_signatures`. This has the added benefit that we don't need to persist the remote `commitment_signed` when we receive it, we can wait for the `tx_signatures` exchange before updating our persisted state.

Reply:
If you want to have a look at the implementation diff for that change, you can find it [here](https://github.com/ACINQ/eclair/pull/2614) for eclair.

Reply:
> Note that the fact that the node didn't remember that channel means it never sent its commitment_signed, which means the other node hasn't sent its tx_signatures and can thus safely forget that channel when receiving the error.

...

> This tx_abort lets the other peer know that they can safely forget that RBF attempt (since we didn't send commitment_signed, they haven't sent tx_signatures so they're not risking anything).

One thing to be careful about is not assuming the peer is honest.  We don't want to forget a channel or RBF attempt *only* because the peer responded with `error` or `tx_abort`.  We also need to verify from our own records that we haven't sent `tx_signatures`.

Reply:
Yes definitely: once you've sent `tx_signatures`, you need to remember the channel until it has been spent or double-spent, regardless of what your peer tells you.

Reply:
Since splicing is another user of `interactive-tx`, reviewers may be interested in our [splicing proposal](https://gist.github.com/t-bast/1ac31f4e27734a10c5b9847d06db8d86), in which we also use these `channel_reestablish` changes (because without it, a disconnection at the wrong time would lead to a force-close, which isn't acceptable).

Reply:
Here is a patch including the changes proposed: https://github.com/niftynei/lightning-rfc/pull/13

Reply:
ACK.

Love that we don’t need to store commitment signed state in the DB by just resending it.

A few thoughts: you should probably enter the splice commitment process when *sending* next_funding tlv, since signing order could be you first.

Following that you should *not* enter a second commitment process when receiving the next_funding tlv, so you don’t end up doing it twice.

Easy enough reason about thought, not sure it requires explicitly mentioning in the spec.

“next_funding” is a little obscure of a name to me. Something like “retry_splice” would be a much clearer name to me, unless this is being used for other purposes as well.

Lastly we may want to be more explicit about when to consider the splice reestablish-able. I would propose the specific moment we consider it that is when we receive tx_complete.

Reply:
> Easy enough reason about though, not sure it requires explicitly mentioning in the spec.

Agreed, I don't  think this is worth mentioning. At least it's clearly not worth mentioning now, because for dual funding it simply doesn't apply. We can see if it makes sense to add in the splice PR when we make more progress on the splice spec PR.

> “next_funding” is a little obscure of a name to me. Something like “retry_splice” would be a much clearer name to me, unless this is being used for other purposes as well.

But this is used for dual funding as well, so it definitely cannot be named `retry_splice`! It is useful for any `interactive-tx` session, which will also very likely encompass mutual closing in the future, so we want something generic enough. This is for the next transaction spending the current funding output, and for dual funding and splicing it is a funding transaction: this is why I think `next_funding_tx` makes sense. But if we start using this for closing txs, I guess we can only do `next_interactive_tx` to cover all cases, but that's something we can change when we specify interactive-tx-based closing.

> Lastly we may want to be more explicit about when to consider the splice reestablish-able. I would propose the specific moment we consider it that is when we receive tx_complete.

Again, this is the dual funding PR, we shouldn't mention splice, and the use-case it currently solves is unrelated to splice. IMO we should consider the interactive-tx reestablish-able as soon as we send our `commit_sig`, which is what is currently specified.